reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -160,61 +146,159 @@ BoundedScheduledExecutorService createBoundedExecutorService() {
-			if (e != null) {
-				e.shutdownNow();
+		for (;;) {
+			SchedulerState<BoundedServicesState> previous = state;
+
+			if (previous.currentResource.boundedServices == SHUTTING_DOWN) {
+				continue;
+			}
+
+			if (previous.currentResource.boundedServices == SHUTDOWN ) {
+				// A dispose process might be ongoing, but we want a forceful shutdown,
+				// so we do our best to release the resources without updating the state.
+				if (previous.initialResource != null) {
+					previous.initialResource.boundedServices.evictor.shutdownNow();
+					if (previous.initialResource.disposedBoundedStates != null) {
+						for (BoundedState bs : previous.initialResource.disposedBoundedStates) {
+							bs.shutdown(true);
+						}
+					}
+				}
+				return;
+			}
+
+			SchedulerState<BoundedServicesState> shuttingDown = SchedulerState.transition(
+					previous.currentResource, SHUTTING_DOWN_STATE, this
+			);
+			if (STATE.compareAndSet(this, previous, shuttingDown)) {
+				assert shuttingDown.initialResource != null;
+				final BoundedState[] toAwait =
+						shuttingDown.initialResource.boundedServices.dispose(true);
+				shuttingDown.initialResource.boundedServices.evictor.shutdownNow();
+
+				SchedulerState<BoundedServicesState> shutdown = SchedulerState.transition(
+						disposed(shuttingDown.initialResource.boundedServices, toAwait),
+						SHUTDOWN_STATE,
+						this
+				);
+				STATE.lazySet(this, shutdown);
 			}
-			services.dispose();
 		}
 	}
 
+	@Override
+	public Mono<Void> disposeGracefully(Duration gracePeriod) {
+		return Mono.defer(() -> {
+			for (;;) {
+				SchedulerState<BoundedServicesState> previous = state;
+
+				if (previous.currentResource.boundedServices == SHUTTING_DOWN) {
+					// A rival thread is performing the shutdown, keep spinning.
+					continue;
+				}
+
+				if (previous.currentResource.boundedServices == SHUTDOWN) {
+					return previous.onDispose;
+				}
+
+				SchedulerState<BoundedServicesState> shuttingDown =
+						SchedulerState.transition(
+								previous.currentResource, SHUTTING_DOWN_STATE, this
+						);
+
+				if (STATE.compareAndSet(this, previous, shuttingDown)) {
+					// We need an intermediate state to make sure the disposal happens just once
+					// and that we capture the BoundedStates, otherwise we'd lose references to schedulers
+					// that we need to wait for.
+					assert shuttingDown.initialResource != null;
+					final BoundedState[] toAwait =
+							shuttingDown.initialResource.boundedServices.dispose(false);
+					shuttingDown.initialResource.boundedServices.evictor.shutdown();
+					SchedulerState<BoundedServicesState> shutdown =
+							SchedulerState.transition(
+									disposed(shuttingDown.initialResource.boundedServices, toAwait),
+									SHUTDOWN_STATE,
+									this
+							);
+					STATE.lazySet(this, shutdown);
+					return shutdown.onDispose;
+				} // else loop until the winner sets SHUTDOWN
+			}
+		}).timeout(gracePeriod);
+	}
```

Comment:
```
@Override
	public void dispose() {
		final SchedulerState<BoundedServicesState> previous = state;

		if (previous.currentResource.boundedServices == SHUTDOWN ) {
			// A dispose process might be ongoing, but we want a forceful shutdown,
			// so we do our best to release the resources without updating the state.
			if (previous.initialResource != null) {
				previous.initialResource.boundedServices.evictor.shutdownNow();
				if (previous.initialResource.disposedBoundedStates != null) {
					for (BoundedState bs : previous.initialResource.disposedBoundedStates) {
						bs.shutdown(true);
					}
				}
			}
			return;
		}

		final BoundedState[] toAwait =
				previous.currentResource.boundedServices.dispose(true);

		SchedulerState<BoundedServicesState> shuttingDown = SchedulerState.transition(
				previous.currentResource, SHUTTING_DOWN_STATE, this
		);
		if (STATE.compareAndSet(this, previous, shuttingDown)) {
			assert shuttingDown.initialResource != null;
			shuttingDown.initialResource.boundedServices.evictor.shutdownNow();

			SchedulerState<BoundedServicesState> shutdown = SchedulerState.transition(
					disposed(shuttingDown.initialResource.boundedServices, toAwait),
					SHUTDOWN_STATE,
					this
			);
			STATE.lazySet(this, shutdown);
		}
	}

	@Override
	public Mono<Void> disposeGracefully(Duration gracePeriod) {
		return Mono.defer(() -> {
			SchedulerState<BoundedServicesState> previous = state;

			if (previous.currentResource.boundedServices == SHUTDOWN) {
				return previous.onDispose;
			}

			final BoundedState[] toAwait =
					previous.currentResource.boundedServices.dispose(false);
			
			if (toAwait == BoundedServices.ALL_SHUTDOWN) {
				return Mono.empty();
			}
			
			previous.currentResource.boundedServices.evictor.shutdown();
			final SchedulerState<BoundedServicesState> shutdown =
					SchedulerState.transition(
							disposed(previous.currentResource.boundedServices, toAwait),
							SHUTDOWN_STATE,
							this
					);

			if (STATE.compareAndSet(this, previous, shutdown)) {
				return shutdown.onDispose;
			}

			// dispose concurrently set state but likely did not observe 
			// BoundedStates, so we need to force terminate it here
			for (BoundedState boundedState : toAwait) {
				boundedState.shutdown(true);
			}
			
			return Mono.empty();
		}).timeout(gracePeriod);
	}
	```

how about this as alternative solution

Reply:
This looks interesting. No busy spinning. The trick with `boundedStates` disposal in case of a race sounds tempting, but no longer delivers the promise of graceful disposal within a given grace period. Another drawback of this approach seems to be the fact that the `dispose()` method would invalidate a concurrent `start()` by the forceful `lazySet` at the end (probably needs a change to `set` as per your other suggestions, or `compareAndSet` in which case it is probably possible to get rid of the undesired invalidation.
All in all, I feel the `boundedState.shutdown(true);` is against users' expectations.

Reply:
Also, if there is a race, the forceful shutdown of `boundedState`s returns an empty `Mono` and gives no chance to await the disposal of `evictor` nor the actual disposal of the workers - even though they're forcefully shutdown, they might still take some time to release resources.

Reply:
@chemicL I did a mistake, forgot to remove old lazySet from the copypasted code. edited

Reply:
> All in all, I feel the boundedState.shutdown(true); is against users' expectations.

that was also a type, copypasted code